### ISSUE TITLE
secondary controllers registration errors

### ISSUE THREAD

**Issue by senorsmile** [PyratLabs/ansible-role-k3s#106]
### Summary

I am testing with a 6 node cluster in Vagrant. 
3 controller nodes. 
3 agents. 

The first controller comes up okay, but the other 2 time out. 

I went into controller 2, copied the line from /etc/systemd/systemd/k3s.service and ran it and got: 
```#
# /usr/local/bin/k3s server --server https://192.168.85.101:6443 --token-file /etc/rancher/k3s/cluster-token --config /etc/rancher/k3s/config.yaml

INFO[2021-04-08T14:13:15.236433093Z] Starting k3s v1.20.5+k3s1 (355fff30)         
INFO[2021-04-08T14:13:15.277972316Z] Managed etcd cluster not yet initialized     
FATA[2021-04-08T14:13:15.288235958Z] starting kubernetes: preparing server: start managed database: joining etcd cluster: etcdclient: no available endpoints
```

on controller1, I see etcd is up and advertising: 
```#
# ss -lnp | grep 2379

tcp                LISTEN              0                    4096                                                                                      127.0.0.1:2379                                              0.0.0.0:*                      users:(("k3s-server",pid=8614,fd=10))                                          
tcp                LISTEN              0                    4096                                                                                      10.0.2.15:2379                                              0.0.0.0:*                      users:(("k3s-server",pid=8614,fd=9))                                           
```

### Issue Type

- Bug Report

### Controller Environment and Configuration

```#yml
# Begin ANSIBLE VERSION
ansible 2.9.19
  config file = None
  configured module search path = ['/home/senorsmile/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/senorsmile/.local/share/virtualenvs/k3s-pyratlabs-it8gg1QU/lib/python3.6/site-packages/ansible
  executable location = /home/senorsmile/.local/share/virtualenvs/k3s-pyratlabs-it8gg1QU/bin/ansible
  python version = 3.6.12 (default, Feb 14 2021, 11:31:45) [GCC 10.2.0]
# End ANSIBLE VERSION

# Begin ANSIBLE CONFIG
DEFAULT_CALLBACK_WHITELIST(env: ANSIBLE_CALLBACK_WHITELIST) = ['timer', 'profile_tasks']
DEFAULT_STDOUT_CALLBACK(env: ANSIBLE_STDOUT_CALLBACK) = yaml
# End ANSIBLE CONFIG

# Begin ANSIBLE ROLES
usage: ansible-galaxy [-h] [--version] [-v] TYPE ...

Perform various Role and Collection related operations.

positional arguments:
  TYPE
    collection   Manage an Ansible Galaxy collection.
    role         Manage an Ansible Galaxy role.

optional arguments:
  --version      show program's version number, config file location,
                 configured module search path, module location, executable
                 location and exit
  -h, --help     show this help message and exit
  -v, --verbose  verbose mode (-vvv for more, -vvvv to enable connection
                 debugging)
# End ANSIBLE ROLES

# Begin PLAY HOSTS
["vagrant_k3s_master01", "vagrant_k3s_master02", "vagrant_k3s_master03", "vagrant_k3s_node01", "vagrant_k3s_node02"]
# End PLAY HOSTS

# Begin K3S ROLE CONFIG
## vagrant_k3s_master01
k3s_become_for_all: true
k3s_install_hard_links: true
k3s_release_version: "v1.20.5+k3s1"
k3s_state: "installed"
k3s_advertise_address: "192.168.85.101"
k3s_node_external_address: "192.168.85.101"
k3s_node_ip_address: "192.168.85.101"
k3s_registration_address: "192.168.85.101"
k3s_control_node: true
k3s_etcd_datastore: true

## vagrant_k3s_master02
k3s_become_for_all: true
k3s_install_hard_links: true
k3s_release_version: "v1.20.5+k3s1"
k3s_state: "installed"
k3s_advertise_address: "192.168.85.102"
k3s_node_external_address: "192.168.85.102"
k3s_node_ip_address: "192.168.85.102"
k3s_registration_address: "192.168.85.101"
k3s_control_node: true
k3s_etcd_datastore: true

## vagrant_k3s_master03
k3s_become_for_all: true
k3s_install_hard_links: true
k3s_release_version: "v1.20.5+k3s1"
k3s_state: "installed"
k3s_advertise_address: "192.168.85.103"
k3s_node_external_address: "192.168.85.103"
k3s_node_ip_address: "192.168.85.103"
k3s_registration_address: "192.168.85.101"
k3s_control_node: true
k3s_etcd_datastore: true

## vagrant_k3s_node01
k3s_become_for_all: true
k3s_install_hard_links: true
k3s_release_version: "v1.20.5+k3s1"
k3s_state: "installed"
k3s_advertise_address: "192.168.85.111"
k3s_node_external_address: "192.168.85.111"
k3s_node_ip_address: "192.168.85.111"
k3s_registration_address: "192.168.85.101"
k3s_control_node: false

## vagrant_k3s_node02
k3s_become_for_all: true
k3s_install_hard_links: true
k3s_release_version: "v1.20.5+k3s1"
k3s_state: "installed"
k3s_advertise_address: "192.168.85.112"
k3s_node_external_address: "192.168.85.112"
k3s_node_ip_address: "192.168.85.112"
k3s_registration_address: "192.168.85.101"
k3s_control_node: false

# End K3S ROLE CONFIG

# Begin K3S RUNTIME CONFIG
## vagrant_k3s_master01
## vagrant_k3s_master02
## vagrant_k3s_master03
## vagrant_k3s_node01
## vagrant_k3s_node02
# End K3S RUNTIME CONFIG
```

```text
* a3c4d9c - (HEAD -> main, tag: v2.8.1, origin/main, origin/HEAD) Updated CHANGELOG (2 weeks ago) <Xan Manning>
```

### Steps to Reproduce

Here's my inventory
```yaml
---
k3s:
  vars:
    ansible_ssh_common_args: '-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
    ansible_user: vagrant
    ansible_become_pass: vagrant
    ansible_ssh_private_key_file: ".vagrant/machines/{{ inventory_hostname }}/virtualbox/private_key"
   
    k3s_become_for_all: True
    k3s_install_hard_links: True
    k3s_release_version: "v1.20.5+k3s1"
    k3s_state: installed
    k3s_advertise_address: "{{ ansible_host }}"
    k3s_node_external_address: "{{ ansible_host }}"
    k3s_node_ip_address: "{{ ansible_host }}"

    k3s_registration_address: 192.168.85.101


  children:
    k3s_masters:
      vars:
        k3s_control_node: True
        k3s_etcd_datastore: True # embedded replicated etcd
      hosts:
        vagrant_k3s_master01:
          ansible_host: 192.168.85.101
        vagrant_k3s_master02:
          ansible_host: 192.168.85.102
        vagrant_k3s_master03:
          ansible_host: 192.168.85.103

    k3s_nodes:
      vars:
        k3s_control_node: False
      hosts:
        vagrant_k3s_node01:
          ansible_host: 192.168.85.111
        vagrant_k3s_node02:
          ansible_host: 192.168.85.112
        vagrant_k3s_node02:
          ansible_host: 192.168.85.112
```

### Expected Result

all checks should pass because service came up properly.  

### Actual Result

```text
TASK [Pyratlabs.k3s : Check that the control plane is available to accept connections] **********************************************************************
Thursday 08 April 2021  06:48:39 -0700 (0:00:08.346)       0:01:04.897 ******** 
skipping: [vagrant_k3s_node01]
skipping: [vagrant_k3s_node02]
ok: [vagrant_k3s_master01]
fatal: [vagrant_k3s_master02]: FAILED! => changed=false 
  elapsed: 300
  msg: Timeout when waiting for 127.0.0.1:6443
fatal: [vagrant_k3s_master03]: FAILED! => changed=false 
  elapsed: 300
  msg: Timeout when waiting for 127.0.0.1:6443
[WARNING]: flush_handlers task does not support when conditional
```

**Comment by xanmanning** [comment]
Invalid configuration variables...

```yaml
    k3s_advertise_address: "{{ ansible_host }}"
    k3s_node_external_address: "{{ ansible_host }}"
    k3s_node_ip_address: "{{ ansible_host }}"
```

These should be set under `k3s_server` and `k3s_agent` dictionary

Here's an example, sorry tt's a bit scrappy, this is a personal project I have been mucking about with: https://github.com/xanmanning/vagrant-kubernetes-cluster/blob/master/playbooks/kube-bootstrap.yml#L20-L43

**Comment by senorsmile** [comment]
> These should be set under k3s_server and k3s_agent dictionary
Ohhh!  they are subkeys to e.g. the k3s_server dict.  Cool.  I'll change that and retry.

**Comment by senorsmile** [comment]
I've found a new issue that I believe is in fact a bug.  I'm closing this and opening a new issue.

**Comment by senorsmile** [comment]
Since I'm still having issues, I'm reopening this.  

None of the agent nodes are connecting, although there are no errors in the ansible run.  

My guess is that, for some reason, it's not passing the `--server IP` to k3s.  I've tried tracing this out but I'm running into a little too much misdirection to easily find how it's being called.  

My inventory now looks like: 

```#
---
k3s:
  vars:
    ansible_ssh_common_args: '-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
    ansible_user: vagrant
    ansible_become_pass: vagrant
    ansible_ssh_private_key_file: ".vagrant/machines/{{ inventory_hostname }}/virtualbox/private_key"
   
    k3s_release_version: "v1.20.5+k3s1"
    k3s_registration_address: 192.168.85.101
    k3s_become_for_all: True
    k3s_install_hard_links: True
    k3s_state: installed
    k3s_use_experimental: True


  children:
    #-------------------
    k3s_masters:
    #-------------------
      vars:
        k3s_control_node: True
        k3s_etcd_datastore: True # embedded replicated etcd
        k3s_server:
          advertise-address: "{{ ansible_host }}"
          node-external-ip: "{{ ansible_host }}"
          node-ip: "{{ ansible_host }}"
          #node-taint:
          #    - "k3s-controlplane=true:NoExecute"

      hosts:
        vagrant-k3s-master01:
          ansible_host: 192.168.85.101
          k3s_primary_control_node: True
        vagrant-k3s-master02:
          ansible_host: 192.168.85.102
        vagrant-k3s-master03:
          ansible_host: 192.168.85.103


    #-------------------
    k3s_nodes:
    #-------------------
      vars:
        k3s_control_node: False
        k3s_agent:
          advertise-address: "{{ ansible_host }}"
          node-external-ip: "{{ ansible_host }}"
          node-ip: "{{ ansible_host }}"
          with-node-id: true

      hosts:
        vagrant-k3s-node01:
          ansible_host: 192.168.85.111
        vagrant-k3s-node02:
          ansible_host: 192.168.85.112---
k3s:
  vars:
    ansible_ssh_common_args: '-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
    ansible_user: vagrant
    ansible_become_pass: vagrant
    ansible_ssh_private_key_file: ".vagrant/machines/{{ inventory_hostname }}/virtualbox/private_key"
   
    k3s_release_version: "v1.20.5+k3s1"
    k3s_registration_address: 192.168.85.101
    k3s_become_for_all: True
    k3s_install_hard_links: True
    k3s_state: installed
    k3s_use_experimental: True


  children:
    #-------------------
    k3s_masters:
    #-------------------
      vars:
        k3s_control_node: True
        k3s_etcd_datastore: True # embedded replicated etcd
        k3s_server:
          advertise-address: "{{ ansible_host }}"
          node-external-ip: "{{ ansible_host }}"
          node-ip: "{{ ansible_host }}"
          #node-taint:
          #    - "k3s-controlplane=true:NoExecute"

      hosts:
        vagrant-k3s-master01:
          ansible_host: 192.168.85.101
          k3s_primary_control_node: True
        vagrant-k3s-master02:
          ansible_host: 192.168.85.102
        vagrant-k3s-master03:
          ansible_host: 192.168.85.103


    #-------------------
    k3s_nodes:
    #-------------------
      vars:
        k3s_control_node: False
        k3s_agent:
          advertise-address: "{{ ansible_host }}"
          node-external-ip: "{{ ansible_host }}"
          node-ip: "{{ ansible_host }}"
          with-node-id: true

      hosts:
        vagrant-k3s-node01:
          ansible_host: 192.168.85.111
        vagrant-k3s-node02:
          ansible_host: 192.168.85.112
        vagrant-k3s-node03:
          ansible_host: 192.168.85.113

        vagrant-k3s-node03:
          ansible_host: 192.168.85.113
```

**Comment by xanmanning** [comment]
I can't recreate this:

https://asciinema.org/a/Flh0UScmcmWOMnDjvvIutttPl

My files: [ansible-role-k3s_issue-106.tar.gz](https://github.com/PyratLabs/ansible-role-k3s/files/6290627/ansible-role-k3s_issue-106.tar.gz)

**Comment by senorsmile** [comment]
I destroyed all of the vm's and tried again from scratch with the same thing happening.  

I have pushed my working dir to github: https://github.com/senorsmile/vagrant-pyratlabs-k3s-ansible. 
I'll continue looking around in case you don't have time to look through things or try it out.

**Comment by xanmanning** [comment]
I had a go with your codebase - I did have to install ansible 2.9.19 into a virtualenv because I can't get pipenv to work on my laptop (no idea). It still seems to work, however I am on Python 3.8 which is the only fundamental difference. I did push an update yesterday to address #105, thought I don't think it would affect your code as you're using 2.9.19 - might be worth updating your git submodule anyway?

https://asciinema.org/a/9mBq8CEW4yu2Q8zI7V8MlNz9l

I'm a bit stuck for ideas otherwise, my laptop is native Linux rather than Windows/WSL so I am not able to run tests against a similar environment, I'd need to spin something up in Azure where I can do some nested virtualisation - might need to wait on that as my cloud bills are a little high this month.

**Comment by senorsmile** [comment]
The Vagrantfile I'm using is a copied template with too much unused code.  I'm on Manjaro at the moment.  I have a build server and another laptop I can try this on.  

I'll use asdf to use python 3.8, update the submodule and try on a different laptop.  

Thanks for all the assistance so far.

**Comment by senorsmile** [comment]
I took another look at the terminal recording and noticed that you are in fact showing the same problem that I do now.  
I reread my last message and somehow forgot to mention the current problem. 

My Vagrantfile has 6 vm's, 3 control nodes and 3 agents.  
When you run `kubectl get nodes` at the end, it only shows the 3 control nodes.  The 3 agent nodes are not joining the cluster properly.

**Comment by xanmanning** [comment]
Remove this line :slightly_smiling_face: : https://github.com/senorsmile/vagrant-pyratlabs-k3s-ansible/commit/2bdeb374e64b70e7f9c45d0271e0da127ce017fb#diff-52c4d436077466b6568ae858b530b03b36099622b78d3518cd50776b9699a64aR47

**Comment by senorsmile** [comment]
Confirmed!  I misinterpreted the advertise address to be the initial server join address... or maybe I was just completely confused!  
Thanks!
